### PR TITLE
Rename callbacks/base.py to callbacks/callback.py

### LINKED ISSUE #11973: Rename `callbacks/base.py` to `callbacks/callback.py`
## Proposed refactor

Rename `callbacks/base.py` to `callbacks/callback.py`

### Motivation
For consistency, let's conform to the standard of our other main files, like `accelerators/accelerator.py` and `strategies/strategy.py`.

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/11908

cc @borda @justusschock @awaelchli @rohitgr7 @ananthsub @daniellepintz

### PR DESCRIPTION
**Pull request by shantam-8**
## What does this PR do?
Renames callbacks/base.py to callbacks/callback.py and changes other instances in different parts of the repository. Also changes callbacks/base.py to provide a deprecation warning.

Fixes #11973

**IMPORTANT: This PR needs to be "Rebase and Merge"d. Do not add merge commits!**

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/base.py [hunk 1]
```diff
@@ -11,347 +11,14 @@
...
-        """Called after ``loss.backward()`` and before optimizers are stepped."""
-
-    def on_before_optimizer_step(
-        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", optimizer: Optimizer, opt_idx: int
-    ) -> None:
-        """Called before ``optimizer.step()``."""
-
-    def on_before_zero_grad(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", optimizer: Optimizer) -> None:
-        """Called before ``optimizer.zero_grad()``."""
+class DDPPlugin(DDPStrategy):
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/base.py:18]: is this copy/pasted? this is the incorrect deprecation warning for Callback

**Review by ananthsub — CHANGES_REQUESTED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/callbacks/base.py [hunk 2]
```diff
@@ -11,347 +11,17 @@
...
-        """Called after ``loss.backward()`` and before optimizers are stepped."""
-
-    def on_before_optimizer_step(
-        self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", optimizer: Optimizer, opt_idx: int
-    ) -> None:
-        """Called before ``optimizer.step()``."""
-
-    def on_before_zero_grad(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", optimizer: Optimizer) -> None:
-        """Called before ``optimizer.zero_grad()``."""
+class Callback(NewCallback):
```
Comment (carmocca) [inline pytorch_lightning/callbacks/base.py:20]: @rohitgr7 Can we follow the same approach as in https://github.com/PyTorchLightning/pytorch-lightning/pull/12740?

Otherwise git blaming the past will have this file jump.

If backwards compatibility with master is a concern, we could clean up the commits and then request Lexie to "rebase and merge" the PR

Comment (rohitgr7) [reply pytorch_lightning/callbacks/base.py:20]: mind check?

Comment (carmocca) [reply pytorch_lightning/callbacks/base.py:20]: LGTM. Reworded the second commit!

Comment (carmocca) [reply pytorch_lightning/callbacks/base.py:20]: Also included the CHANGELOG entry in the second commit

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@rohitgr7 could you rebase once more? Thx

**Review by carmocca — APPROVED** [review]
